hyperlink/test/test_decoded_url.py

```diff
@@ -0,0 +1,139 @@
+        assert bdurl.scheme == 'http'
+        assert bdurl.host == 'example.com'
+        assert bdurl.port == 80
+        assert bdurl.path == ('',)
+        assert bdurl.fragment == ''
+
+        durl = DecodedURL.from_text(TOTAL_URL)
+
+        assert durl.scheme == 'https'
+        assert durl.host == 'bücher.ch'
+        assert durl.port == 8080
+        assert durl.path == ('a', 'nice nice', '.', 'path', '')
+        assert durl.fragment == 'frég'
+        assert durl.get('zot') == ['23%']
+
+        assert durl.user == 'user'
+        assert durl.userinfo == ('user', '\0\0\0\0')
+
+    def test_passthroughs(self):
+        # just basic tests for the methods that more or less pass straight
+        # through to the underlying URL
+
+        durl = DecodedURL.from_text(TOTAL_URL)
+        assert durl.sibling('te%t').path[-1] == 'te%t'
+        assert durl.child('../test2%').path[-1] == '../test2%'
+        assert durl.child() == durl
+        assert durl.child() is durl
+        assert durl.click('/').path[-1] == ''
+        assert durl.user == 'user'
+
+        assert '.' in durl.path
+        assert '.' not in durl.normalize().path
+
+        assert durl.to_uri().fragment == 'fr%C3%A9g'
+        assert ' ' in durl.to_iri().path[1]
+
+        assert durl.to_text(with_password=True) == TOTAL_URL
+
+        assert durl.absolute
+        assert durl.rooted
+
+        assert durl == durl.encoded_url.get_decoded_url()
+
+        durl2 = DecodedURL.from_text(TOTAL_URL, lazy=True)
+        assert durl2 == durl2.encoded_url.get_decoded_url(lazy=True)
+
+        # TODO change this to the actual value when str() changes get merged
+        assert '%20' in str(DecodedURL.from_text(BASIC_URL).child(' '))
+
+        assert not (durl == 1)
+        assert durl != 1
+
+    def test_repr(self):
+        durl = DecodedURL.from_text(TOTAL_URL)
+        assert repr(durl) == 'DecodedURL(url=' + repr(durl._url) + ')'
+
+    def test_query_manipulation(self):
+        durl = DecodedURL.from_text(TOTAL_URL)
+
+        assert durl.get('zot') == ['23%']
+        durl = durl.add(' ', 'space')
+        assert durl.get(' ') == ['space']
+        durl = durl.set(' ', 'spa%ed')
+        assert durl.get(' ') == ['spa%ed']
+
+        durl = DecodedURL(url=durl.to_uri())
+        assert durl.get(' ') == ['spa%ed']
+        durl = durl.remove(' ')
+        assert durl.get(' ') == []
+
+        durl = DecodedURL.from_text('/?%61rg=b&arg=c')
+        assert durl.get('arg') == ['b', 'c']
+
+        assert durl.set('arg', 'd').get('arg') == ['d']
+
+    def test_equivalences(self):
+        durl = DecodedURL.from_text(TOTAL_URL)
+        durl2 = DecodedURL.from_text(TOTAL_URL)
+        burl = DecodedURL.from_text(BASIC_URL)
+
+        assert durl == durl
+        assert durl == durl2
+        assert durl != burl
+        assert durl != None
+        assert durl != durl._url
+
+        durl_map = {}
+        durl_map[durl] = durl
+        durl_map[durl2] = durl2
```

Comment:
This should also test that a `URL` and a `DecodedURL` that represent the same underlying URL don't overlap in a `dict` (or `set`.)